File: native-host/clipboard_helper.swift
import AppKit
import Foundation

func readMessage() -> [String: Any]? {
    var lengthBytes = [UInt8](repeating: 0, count: 4)
    guard fread(&lengthBytes, 1, 4, stdin) == 4 else { return nil }
    let length = UInt32(lengthBytes[0])
        | UInt32(lengthBytes[1]) << 8
        | UInt32(lengthBytes[2]) << 16
        | UInt32(lengthBytes[3]) << 24
    guard length > 0, length < 50_000_000 else { return nil }

    var buffer = [UInt8](repeating: 0, count: Int(length))
    guard fread(&buffer, 1, Int(length), stdin) == Int(length) else { return nil }

    return try? JSONSerialization.jsonObject(with: Data(buffer)) as? [String: Any]
}

func sendMessage(_ dict: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: dict) else { return }
    var length = UInt32(data.count)
    fwrite(&length, 4, 1, stdout)
    _ = data.withUnsafeBytes { fwrite($0.baseAddress!, 1, data.count, stdout) }
    fflush(stdout)
}

func copyGifToClipboard(base64: String) -> Bool {
    guard let data = Data(base64Encoded: base64) else { return false }

    let gifType = NSPasteboard.PasteboardType("com.compuserve.gif")
    let pb = NSPasteboard.general
    pb.clearContents()

    if let image = NSImage(data: data), let tiff = image.tiffRepresentation {
        pb.declareTypes([gifType, .tiff], owner: nil)
        pb.setData(data, forType: gifType)
        pb.setData(tiff, forType: .tiff)
    } else {
        pb.declareTypes([gifType], owner: nil)
        pb.setData(data, forType: gifType)
    }

    return true
}

guard let msg = readMessage(), let action = msg["action"] as? String else {
    sendMessage(["success": false, "error": "Invalid message"])
    exit(1)
}

if action == "copyGif", let base64 = msg["base64"] as? String {
    sendMessage(["success": copyGifToClipboard(base64: base64)])
} else {
    sendMessage(["success": false, "error": "Unknown action or missing data"])
}

File: native-host/clipboard_helper.py
#!/usr/bin/env python3
import json
import struct
import sys
import base64

import AppKit


def read_message():
    raw = sys.stdin.buffer.read(4)
    if len(raw) < 4:
        return None
    length = struct.unpack("<I", raw)[0]
    if length == 0 or length > 50_000_000:
        return None
    data = sys.stdin.buffer.read(length)
    if len(data) < length:
        return None
    return json.loads(data)


def send_message(obj):
    data = json.dumps(obj).encode("utf-8")
    sys.stdout.buffer.write(struct.pack("<I", len(data)))
    sys.stdout.buffer.write(data)
    sys.stdout.buffer.flush()


def copy_gif_to_clipboard(b64):
    data = base64.b64decode(b64)

    gif_type = AppKit.NSPasteboardType("com.compuserve.gif")
    pb = AppKit.NSPasteboard.generalPasteboard()
    pb.clearContents()

    ns_data = AppKit.NSData.dataWithBytes_length_(data, len(data))
    image = AppKit.NSImage.alloc().initWithData_(ns_data)

    types = [gif_type]
    if image and image.TIFFRepresentation():
        types.append(AppKit.NSPasteboardTypeTIFF)

    pb.declareTypes_owner_(types, None)
    pb.setData_forType_(ns_data, gif_type)

    if image and image.TIFFRepresentation():
        pb.setData_forType_(image.TIFFRepresentation(), AppKit.NSPasteboardTypeTIFF)

    return True


msg = read_message()
if not msg or "action" not in msg:
    send_message({"success": False, "error": "Invalid message"})
    sys.exit(1)

if msg["action"] == "copyGif" and "base64" in msg:
    try:
        ok = copy_gif_to_clipboard(msg["base64"])
        send_message({"success": ok})
    except Exception as e:
        send_message({"success": False, "error": str(e)})
else:
    send_message({"success": False, "error": "Unknown action or missing data"})
